unixfs/src/pb.rs

```diff
@@ -0,0 +1,233 @@
+            Some(Cow::Borrowed(_)) | None => return Err(UnixFsReadFailed::NoData),
+        })?;
+
+        Ok(FlatUnixFs { links, data: inner })
+    }
+}
+
+#[cfg(test)]
+impl<'a> FlatUnixFs<'a> {
+    pub fn range_links(&'a self) -> impl Iterator<Item = (PBLink<'a>, Range<u64>)> {
+        assert_eq!(self.links.len(), self.data.blocksizes.len());
+
+        let zipped = self
+            .links
+            .clone()
+            .into_iter()
+            .zip(self.data.blocksizes.iter().copied());
+
+        // important: we have validated links.len() == blocksizes.len()
+        RangeLinks::from_links_and_blocksizes(zipped, Some(0))
+    }
+}
+
+pub(crate) struct RangeLinks<I> {
+    inner: I,
+    base: u64,
+}
+
+impl<'a, I> RangeLinks<I>
+where
+    I: Iterator<Item = (PBLink<'a>, u64)>,
+{
+    /// `start_offset` is the offset of the current tree when walking the graph.
+    pub fn from_links_and_blocksizes(zipped: I, start_offset: Option<u64>) -> RangeLinks<I> {
+        RangeLinks {
+            inner: zipped,
+            base: start_offset.unwrap_or(0),
+        }
+    }
+}
+
+impl<'a, I> Iterator for RangeLinks<I>
+where
+    I: Iterator<Item = (PBLink<'a>, u64)>,
+{
+    type Item = (PBLink<'a>, Range<u64>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.inner.next() {
+            Some((link, blocksize)) => {
+                let returned_base = self.base;
+                self.base += blocksize;
+                Some((link, returned_base..(returned_base + blocksize)))
+            }
+            None => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use super::{FlatUnixFs, PBNode, UnixFs, UnixFsType};
+    use hex_literal::hex;
+    use std::borrow::Cow;
+    use std::convert::TryFrom;
+    use std::fmt;
+
+    #[test]
+    fn hello_world() {
+        use quick_protobuf::{BytesReader, MessageRead};
+        let input = hex!("0a0d08021207636f6e74656e741807");
+
+        let mut reader = BytesReader::from_bytes(&input);
+        let dagnode =
+            PBNode::from_reader(&mut reader, &input).expect("parse outer merkledag::PBNode");
+        assert!(dagnode.Links.is_empty());
+        println!("{:?}", dagnode);
+
+        let unixfs_data = UnixFs::try_from(&dagnode).expect("parse inner unixfs::Data");
+        assert_eq!(unixfs_data.Type, UnixFsType::File);
+        assert_eq!(unixfs_data.Data, Some(Cow::Borrowed(&b"content"[..])));
+        println!("{:?}", unixfs_data);
+    }
+
+    #[test]
+    fn linux_tarxz_range_links() {
+        let input = hex!("122b0a2212203822560f945fd3c74522de3448512a7e45cb53f0a9a1e12161da4667531ec12e120018aed4e015122b0a2212208594eb4dd5d67e573d506cd950ac59863b9afb024a590d7fe49b42fbcb44af43120018aed4e015122b0a221220745a70b6cd7ec3e46d16fb15b5e1e5db256f6a7a52d0b359f8f49b242665e17b120018b4e7e8090a1608021888c1a835208080e015208080e0152088c1e809");
```

Comment:
nit: as before, I'd cut it into 2-3 lines with `\`

Reply:
I was counting on rustfmt to do that for me and as it did not, I thought this to be non-important. The protobuf is really difficult to eyeball as well.

Reply:
hex! is a proc macro, it doesn't understand escape + newline + whitespace :)

Reply:
Yeah, those cheeky macros; all good, same with the other case :+1: